src/main/clojure/clara/rules/durability/fressian.clj

```diff
@@ -191,6 +191,40 @@
                       (read-record add-node-expr-fn)
                       d/cache-node)))}}))
 
+(defn- create-identity-based-handler
+  [clazz
+   tag
+   read-fn
+   write-fn]
+  (let [indexed-tag (str tag "-idx")]
+    ;; Write an object a single time per object reference to that object.  The object is then "cached"
+    ;; with the IdentityHashMap `d/clj-struct-holder`.  If another reference to this object instance
+    ;; is encountered later, only the "index" of the object in the map will be written.
+    {:class clazz
+     :writer (reify WriteHandler
+               (write [_ w o]
+                 (if-let [idx (d/clj-struct->idx o)]
+                   (do
+                     (.writeTag w indexed-tag 1)
+                     (.writeInt w idx))
+                   (do
+                     (write-fn w tag o)
+                     (d/clj-struct-holder-add-fact-idx! o)))))
```

Comment:
I think it is important here that the object is added to the cache _after_ it has been written (and implicitly after everything contained within it is written and cached) to match the ordering of the addition of items to the cache during read.  That is, we wouldn't want a write of 
````
[#{:test}]
````
To result in a cache of 
````
{[#{test}] 0]
  #{test} 1}
````
on write and a cache of 
````
[#{test} [#{test}]]
````
on read.

I ran the following test in my REPL:

````
(defn serde1 [x]
  (with-open [os (java.io.ByteArrayOutputStream.)
              ^FressianWriter wtr (fres/create-writer os :handlers df/write-handler-lookup)]
    ;; Write
    (pform/thread-local-binding [d/node-id->node-cache (volatile! {})
                                 d/clj-struct-holder (java.util.IdentityHashMap.)]
                                (fres/write-object wtr x)
                                (println "Write: " (.get d/clj-struct-holder))

                                )
    ;; Read
    (let [data (.toByteArray os)]
      (pform/thread-local-binding [d/clj-struct-holder (java.util.ArrayList.)]
                                  (with-open [is (java.io.ByteArrayInputStream. data)
                                              ^FressianReader rdr (fres/create-reader is :handlers df/read-handler-lookup)]
                                    (fres/read-object rdr)
                                    (println "Read" (.get d/clj-struct-holder)))))))
````
````
clara.test-fressian> (serde1 [[:a]])
Write:  #object[java.util.IdentityHashMap 0x71cdad81 {[[:a]]=1, [:a]=0}]
Read #object[java.util.ArrayList 0x36758cef [[:a], [[:a]]]]
````

This isn't a new issue, but it is a bigger one if we're doing caching on all Clojure data structures.  I suggest adding a comment here.  Thoughts/disagreements @mrrodriguez ?

Reply:
@WilliamParker I believe you are wanting just to make it clearer in the docs/comments that the write traversal is carefully chosen so that it can be mirrored by the read traversal right?

If so, yeah, I think it should be explained better to avoid bad implementations being written. I also don't think that is something unique to this PR though. It's really just a call for better documentation around the way this works in general right?

Reply:
Right, I don't think it needs to be a blocker for merging this.  It is just a bigger issue now that it isn't just Clojure records being cached.

Reply:
added a minor comment